### PR TITLE
Collision based search

### PR DESCRIPTION
**Pull request by johannschwabe**
Slowest boy

### CODE REVIEW

File: API/Generator/Generator.py [hunk 1]
```diff
@@ -134,8 +135,11 @@ def simulate(self):
 
     @staticmethod
     def creation_ticks(duration, total) -> List[int]:
-        res = []
-        for _ in range(total):
-            res.append(random.randint(0, duration - 1))
-
-        return res
+        return [1] * total
```
Comment (joelbarmettlerUZH) [inline API/Generator/Generator.py]: seems illegal

File: Demos/CBS/Allocator/CBSAllocator.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
+"""
+Python implementation of Conflict-based search
+author: Ashwin Bose (@atb033)
```
Comment (joelbarmettlerUZH) [inline Demos/CBS/Allocator/CBSAllocator.py]: Haste den ganzen code von ihm? Sonst würde ich evt. noch nen kommentar machen das du den Code adaptiert hast, sonst istes gelogen

File: Demos/CBS/Allocator/CBSAllocator.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                    while len(solution[_agent]) > segment_index and solution[_agent][segment_index].min.t <= t:
+                        if solution[_agent][segment_index].max.t >= t:
+                            segment = solution[_agent][segment_index]
+                            posi = segment.coordinates[t - segment.min.t]
+                            bl = posi - new_max_near_radius
+                            tr = posi + new_max_near_radius
+                            intersections: Iterator[int] = tree.intersection(bl.list_rep() + tr.list_rep())
+                            intersecting_agents = set(
+                                [agents[agent_hash] for agent_hash in intersections if agent_hash != hash(_agent)])
+                            for intersecting_agent in intersecting_agents:
```
Comment (joelbarmettlerUZH) [inline Demos/CBS/Allocator/CBSAllocator.py:246]: Krass performance heavy shit right here

File: Demos/CBS/Allocator/CBSCostFunctions.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    def __call__(node: "HighLevelNode"):
+        pass
+
+
+class PathLength(CostFunction):
+    failed_allocation_valid = False
+
+    @staticmethod
+    def __call__(node: "HighLevelNode"):
+        return sum([sum([len(path.coordinates) for path in paths]) for paths in node.solution.values()])
```
Comment (joelbarmettlerUZH) [inline Demos/CBS/Allocator/CBSCostFunctions.py]: Könntest du das ausschreiben? Ist etwas viel Logik für nen one-liner

**Review by joelbarmettlerUZH — COMMENTED** [review]
- Type annotations missing on many functions / methods

File: Demos/CBS/Allocator/CBSCostFunctions.py (hunk 4 above)
Comment (johannschwabe) [reply Demos/CBS/Allocator/CBSCostFunctions.py]: das vom herr mapmapreduce?

File: Demos/CBS/Allocator/CBSAllocator.py (hunk 3 above)
Comment (johannschwabe) [reply Demos/CBS/Allocator/CBSAllocator.py:246]: jap, mir ist aber nix schnelleres eingefallen, um die erste collision zu finden. Hier schreibe ich ja einfach jede allocatete position in einen Rtree in reihenfolge der ticks.

File: API/Generator/Generator.py [hunk 5]
```diff
@@ -134,6 +136,7 @@ def simulate(self):
 
     @staticmethod
     def creation_ticks(duration, total) -> List[int]:
+        # return [1] * total
```
Comment (thommann) [inline API/Generator/Generator.py]: remove

File: API/config.py [hunk 6]
```diff
@@ -1,4 +1,5 @@
+from Demos.CBS.Allocator.CBSAllocator import CBS
 from Demos.FCFS.Allocator.FCFSAllocator import FCFSAllocator
 from Demos.Priority.Allocator.PriorityAllocator import PriorityAllocator
 
-available_allocators = [PriorityAllocator, FCFSAllocator]
+available_allocators = [PriorityAllocator, FCFSAllocator, CBS]
```
Comment (thommann) [inline API/config.py]: Call it CBSAllocator

File: Demos/CBS/Allocator/CBSAllocator.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+
+from rtree import Index
+from rtree.index import Property
+
+from Simulator.Agents.PathAgent import PathAgent
+from Simulator.Allocations.Allocation import Allocation
+from Simulator.Allocations.AllocationHistory import AllocationHistory
+from Simulator.Allocations.AllocationReason import AllocationReason
+from Simulator.Mechanism.Allocator import Allocator
+from Simulator.Segments.PathSegment import PathSegment
```
Comment (thommann) [inline Demos/CBS/Allocator/CBSAllocator.py]: you can import everything directly from Simulator

File: Demos/CBS/Allocator/CBSAllocator.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
+"""
+Implementation of Conflict-based search based on Ashwin Bose (@atb033)'s implementation
+"""
```
Comment (thommann) [inline Demos/CBS/Allocator/CBSAllocator.py]: Do you have a link to the resource?

File: Demos/CBS/Allocator/CBSAllocatorHelpers.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    """
+    HighLevelNode for CBS. Saves a (possibly invalid) allocation for all agents and
+    the constraints under which it was computed
+    """
+
+    def __init__(self):
+        self.solution: Dict["PathAgent", List["PathSegment"]] = dict()
+        self.constraint_dict: Dict["PathAgent", Set["Coordinate4D"]] = dict()
+        self.first_conflict: "Conflict|None" = None
+        self.newly_constraint: "PathAgent|None" = None
```
Comment (thommann) [inline Demos/CBS/Allocator/CBSAllocatorHelpers.py]: use Optional[] from typing instead

File: Demos/CBS/Allocator/CBSAllocatorHelpers.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    def copy(self):
+        cpy = HighLevelNode()
+        for agent in list(self.constraint_dict.keys()):
+            cpy.constraint_dict[agent] = self.constraint_dict[agent].copy()
+        for agent in list(self.solution.keys()):
+            cpy.solution[agent] = self.solution[agent].copy()
+        return cpy
+
+    def __str__(self):
+        res = "-------------------------------------------------------\n"
```
Comment (thommann) [inline Demos/CBS/Allocator/CBSAllocatorHelpers.py]: remove debugging string

File: Demos/CBS/Allocator/CBSCostFunctions.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    @abstractmethod
+    def __call__(node: "HighLevelNode") -> float:
+        pass
+
+
+class PathLength(CostFunction):
+    """
+    Minimize the total number of allocated ticks
+    """
+    failed_allocation_valid = False
```
Comment (thommann) [inline Demos/CBS/Allocator/CBSCostFunctions.py:25]: will this always have a solution?

File: Demos/CBS/BidTracker/CBSBidTracker.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    """
+
+    def __init__(self):
+        """
+        Initialize dict of bids by tick and agent.
+        """
+        self.past_bids: Dict[int, Dict["Agent", List[Union["CBSPathBid", "CBSSpaceBid"]]]] = {}
+
+    def request_new_bid(self, tick: int, agent: "Agent",
+                        environment: "Environment") -> Optional[Union["CBSPathBid", "CBSSpaceBid"]]:
```
Comment (thommann) [inline Demos/CBS/BidTracker/CBSBidTracker.py:22]: are spaceBids possible for CBS?

File: Demos/CBS/CBSAstar/CBSAstar.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+                valid = is_valid_for_path_allocation(self.environment, next_neighbor, agent, constraints)
+                if valid and next_neighbor.t <= self.environment.dimension.t:
+                    neighbor = Node(next_neighbor, current_node, set())
+
+                    # Closed node
+                    if hash(neighbor) in closed_nodes:
+                        continue
+
+                    neighbor.g = current_node.g + 1
+                    # neighbor.h = (end_node.position - neighbor.position).l1
```
Comment (thommann) [inline Demos/CBS/CBSAstar/CBSAstar.py]: remove

File: Demos/CBS/CBSAstar/CBSAstar.py [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+
+        path, steps = self.astar_loop(start, end, agent, constraints)
+
+        if len(path) == 0:
+            print(f"ASTAR failed: {'MaxIter' if steps == self.max_iter else 'No valid Allocation'}")
+            return []
+
+        complete_path = self.complete_path(path, agent)
+
+        # print(f"ASTAR: {complete_path[0]} -> {complete_path[-1]},\tPathLen: {len(path):3d},\tSteps: {steps:3d}")
```
Comment (thommann) [inline Demos/CBS/CBSAstar/CBSAstar.py]: remove

File: Demos/FCFS/Allocator/FCFSAllocator.py [hunk 15]
```diff
@@ -145,6 +146,7 @@ def allocate(self, agents: List["Agent"], environment: "Environment", tick: int)
         """
         astar = AStar(environment, self.bid_tracker, tick)
         allocations: Dict["Agent", "Allocation"] = {}
+        random.shuffle(agents)
```
Comment (thommann) [inline Demos/FCFS/Allocator/FCFSAllocator.py:149]: what is this shuffle for?

File: Simulator/Owners/PathOwner.py [hunk 16]
```diff
@@ -87,7 +87,7 @@ def generate_agents(self, t: int, environment: "Environment") -> List["PathAgent
             stays.pop()
             agent = self.initialize_agent(locations, stays)
             res.append(agent)
-            print(f"{agent} {' -> '.join([str(loc) for loc in locations])}")
+            # print(f"{agent} {' -> '.join([str(loc) for loc in locations])}")
```
Comment (thommann) [inline Simulator/Owners/PathOwner.py]: remove

**Review by thommann — CHANGES_REQUESTED** [review]
I did not find any mistakes in the CBS as far as I can tell. But I don't exactly know how it works so...
I found some minor nitpicks and not so clean parts so I marked those :D

File: Demos/CBS/Allocator/CBSCostFunctions.py (hunk 11 above)
Comment (johannschwabe) [reply Demos/CBS/Allocator/CBSCostFunctions.py:25]: CBS might not find a solution if failed_allocation_valid if False.
Is this what you mean?

File: Demos/CBS/BidTracker/CBSBidTracker.py (hunk 12 above)
Comment (johannschwabe) [reply Demos/CBS/BidTracker/CBSBidTracker.py:22]: Not currently

File: Demos/FCFS/Allocator/FCFSAllocator.py (hunk 15 above)
Comment (johannschwabe) [reply Demos/FCFS/Allocator/FCFSAllocator.py:149]: Agents in FCFS should be allocated in random order and not in the order of their owners

**Pull request merged**